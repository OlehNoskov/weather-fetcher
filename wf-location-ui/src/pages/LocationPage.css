.App {

    .change-location {
        text-align: center;

        .title {
            margin-top: 25px;
            text-align: center;
            font-size: 32px;
            font-family: sans-serif;
        }

        .change-button {
            margin-top: 15px;
            margin-bottom: 15px;
        }
    }
}

.box {
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    width: 550px;
    background-color: #f6f6f6;
    padding: 4px;
    border-radius: 5px;

    .title-modal {
        font-family: sans-serif, sans-serif;
        margin-top: 15px;
        text-align: center;
        font-size: 34px;

        .modal-close-button {
            margin-right: 10px;
            float: right;
            min-width: 0;
            width: 20px;
            height: 20px;
            line-height: 0;
        }

        .close-image {
            height: 20px;
        }
    }

    .inputs {
        display: flex;
        justify-content: center;
        margin-bottom: 30px;
        margin-top: 30px;

        .city {
            margin-left: 20px;
        }
    }

    .alert-message {
        margin-top: 20px;
        text-align: center;
    }

    .search-button {
        text-align: right;
        margin-bottom: 25px;
        margin-right: 30px;
    }
}