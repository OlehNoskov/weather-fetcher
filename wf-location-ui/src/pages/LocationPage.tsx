import React, {useEffect, useState} from "react";

import {Alert, Box, Button, Icon, Modal, TextField, Typography} from "@mui/material";
import {Forecast} from "../dto/Forecast";
import {getForecast} from "../service/Service";
import WeatherForecastPage from "./WeatherForecastPage";

import './LocationPage.css';

export default function LocationPage() {
    const [forecast, setForecast] = useState<Forecast>();
    const [openModal, setOpenModal] = React.useState(false);
    const [country, setCountry] = useState<string>("");
    const [city, setCity] = useState<string>("");

    useEffect(() => {
        window.localStorage.setItem("city", city!);
        window.localStorage.setItem("country", country!);
    }, [country, city]);

    const isDisabledButton = (): boolean => {
        return country?.length <= 2 || city?.length <= 2;
    };

    const getForecastWeather = () => {
        getForecast(country!, city!).then((response) => setForecast(response));


        console.log(getForecast(country!, city!));
        setOpenModal(false);
    };

    return (
        <div className="App">
            <div className={"change-location"}>
                <Typography variant="h6" className={"title"}>
                    Weather forecast
                </Typography>
                {forecast === undefined &&
                    <Alert severity="info" sx={{display: "flex", justifyContent: "center"}}>
                        Please, click on the Change button for getting weather forecast!
                    </Alert>}
                <Button variant="contained" onClick={() => setOpenModal(true)}
                        size={"large"} color="success" className={"change-button"}>
                    Change
                </Button>
            </div>
            <Modal
                open={openModal}
                aria-labelledby="modal-modal-title"
                aria-describedby="modal-modal-description">
                <Box className={"box"}>
                    <Typography variant="h6" className={"title-modal"}>
                        Weather forecast
                        <Button className={"modal-close-button"} onClick={() => setOpenModal(false)}
                                data-testid={"close-modal"}>
                            <Icon>
                                <img className={"close-image"} src="/images/close-icon.svg" alt={'close-icon'}/>
                            </Icon>
                        </Button>
                    </Typography>
                    {isDisabledButton() &&
                        <Alert severity="warning" className={"alert-message"}
                               sx={{display: "flex", justifyContent: "center"}}>
                            Please, input country and city for showing weather forecast!
                        </Alert>}
                    <div className={"inputs"}>
                        <div>
                            <TextField
                                error={isDisabledButton()}
                                label="Country" variant="outlined"
                                type="text" value={country}
                                onChange={(value) => {
                                    setCountry(value.target.value);
                                }} required={true}/>
                        </div>
                        <div>
                            <TextField
                                error={isDisabledButton()}
                                label="City" variant="outlined"
                                className={"city"} type="text" value={city}
                                onChange={(value) => {
                                    setCity(value.target.value);
                                }} required={true}/>
                        </div>
                    </div>
                    <div className={"search-button"}>
                        <Button variant="contained" onClick={getForecastWeather}
                                size={"large"} color="success" disabled={isDisabledButton()}>
                            Search
                        </Button>
                    </div>
                </Box>
            </Modal>
            {forecast !== undefined && <WeatherForecastPage forecast={forecast}/>}
        </div>
    );
};