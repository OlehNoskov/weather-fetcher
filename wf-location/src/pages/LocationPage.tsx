import React, {useEffect, useState} from "react";
import './LocationPage.css';
import {Alert, Box, Button, Icon, Modal, TextField, Typography} from "@mui/material";
import {Forecast} from "../dto/Forecast";
import {getForecast} from "../service/Service";
// import WeatherForecastPage from "./WeatherForecastPage";

export default function LocationPage() {
    // const [forecast, setForecast] = useState<Forecast>();
    // const [openModal, setOpenModal] = React.useState(false);
    // const [country, setCountry] = useState(window.localStorage.getItem("country"));
    // const [city, setCity] = useState(window.localStorage.getItem("city"));
    //
    // const isDisabledButton = (): boolean => {
    //     // @ts-ignore
    //     return country?.length <= 2 || city?.length <= 2;
    // };
    //
    // const isValidCountry = (): boolean => {
    //     // @ts-ignore
    //     return country?.length <= 2;
    // };
    //
    // const isValidCity = (): boolean => {
    //     // @ts-ignore
    //     return city?.length <= 2;
    // };
    //
    // useEffect(() => {
    //     window.localStorage.setItem("country", country!);
    // }, [country]);
    //
    // useEffect(() => {
    //     window.localStorage.setItem("city", city!);
    // }, [city]);
    //
    // const handleOpenModal = () => setOpenModal(true);
    //
    // const handleCloseModal = () => setOpenModal(false);
    //
    // const getForecastWeather = () => {
    //     getForecast(country!, city!).then((response) => setForecast(response));
    //     setOpenModal(false);
    // };
    //
    // function getAlertMessage() {
    //     return isDisabledButton() ?
    //         <Alert severity="warning" className={"alert-message"}>
    //             Please, input country and city for showing weather forecast!
    //         </Alert>
    //         :
    //         <></>
    // }
    //
    // function getWeatherForecast() {
    //     return forecast !== undefined ? <WeatherForecastPage forecast={forecast}/> : <></>
    // }

    return (
        // <div className="App">
        //     <Typography variant="h6" className={"change-location"}>
        //         Change location
        //         <Button variant="contained" onClick={handleOpenModal}
        //                 size={"large"} color="success" className={"change-button"}>Change</Button>
        //     </Typography>
        //     <Modal
        //         open={openModal}
        //         aria-labelledby="modal-modal-title"
        //         aria-describedby="modal-modal-description">
        //         <Box className={"box"}>
        //             <Typography variant="h6" className={"title-modal"}>
        //                 Weather forecast
        //                 <Button className={"modal-close-button"} onClick={handleCloseModal}>
        //                     <Icon>
        //                         <img className={"close-image"} src="/images/close-icon.svg" alt={'close-icon'}/>
        //                     </Icon>
        //                 </Button>
        //             </Typography>
        //             {getAlertMessage()}
        //             <div className={"inputs"}>
        //                 <div>
        //                     <TextField
        //                         error={isValidCountry()}
        //                         label="Country" variant="outlined"
        //                         type="text" value={country}
        //                         onChange={(value) => {
        //                             setCountry(value.target.value);
        //                         }} required={true}/>
        //                 </div>
        //                 <div>
        //                     <TextField
        //                         error={isValidCity()}
        //                         label="City" variant="outlined"
        //                         className={"city"} type="text" value={city}
        //                         onChange={(value) => {
        //                             setCity(value.target.value);
        //                         }} required={true}/>
        //                 </div>
        //             </div>
        //             <div className={"search-button"}>
        //                 <Button variant="contained" onClick={getForecastWeather}
        //                         size={"large"} color="success" disabled={isDisabledButton()}>
        //                     Search
        //                 </Button>
        //             </div>
        //         </Box>
        //     </Modal>
        //     {getWeatherForecast()}
        // </div>
        null
    );
};