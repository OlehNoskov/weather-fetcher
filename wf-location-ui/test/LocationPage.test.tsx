import {screen, render} from '@testing-library/react';
import '@testing-library/jest-dom';
import LocationPage from '../src/pages/LocationPage';
import userEvent from '@testing-library/user-event';

describe('LocationPage', () => {
    const user = userEvent.setup();
    const mockClick = jest.fn();

    global.fetch = mockClick(() => Promise.resolve({
        json: () => Promise.resolve()
    }));

    it('should render LocationPage without opened modal window', () => {
        const {asFragment} = render(<LocationPage/>);

        expect(asFragment()).toMatchSnapshot();
    });

    it('should render LocationPage after click on the button Change', () => {
        const {asFragment} = render(<LocationPage/>);
        const changeButton = screen.getByText('Change');
        user.click(changeButton);

        expect(mockClick).toHaveBeenCalledTimes(1);
        expect(asFragment()).toMatchSnapshot();
    });


    it('should render modal with inputs and disabled button after typing incorrect text', async () => {
        const {asFragment} = render(<LocationPage/>);
        const changeButton = screen.getByText('Change');
        await user.click(changeButton);

        const inputCountry = screen.getByText('Country');
        inputCountry.style.pointerEvents = 'auto';
        expect(inputCountry).not.toHaveFocus();
        await user.type(inputCountry, 'qq');

        const inputCity = screen.getByText('City');
        inputCity.style.pointerEvents = 'auto';
        expect(inputCity).not.toHaveFocus();
        await user.type(inputCity, 'qq');

        const searchButton = screen.getByText('Search');
        searchButton.style.pointerEvents = 'auto';
        await user.click(searchButton);

        expect(searchButton).toHaveProperty('disabled', true);
        expect(asFragment()).toMatchSnapshot();
    });

    it('should render modal with inputs and active button after typing correct text and show weather forecast', async () => {
        // @ts-ignore
        global.fetch = jest.fn(() => Promise.resolve({
            json: () => Promise.resolve()
        }));

        const {asFragment} = render(<LocationPage/>);
        const changeButton = screen.getByText('Change');
        await user.click(changeButton);

        const inputCountry = screen.getByText('Country');
        inputCountry.style.pointerEvents = 'auto';
        expect(inputCountry).not.toHaveFocus();
        await user.type(inputCountry, "Ukraine");

        const inputCity = screen.getByText('City');
        inputCity.style.pointerEvents = 'auto';
        expect(inputCity).not.toHaveFocus();
        await user.type(inputCity, 'Kharkiv');

        const searchButton = screen.getByText('Search');
        searchButton.style.pointerEvents = 'auto';
        await user.click(searchButton);

        expect(searchButton).toHaveProperty('disabled', false);
        expect(asFragment()).toMatchSnapshot();
    });
});
