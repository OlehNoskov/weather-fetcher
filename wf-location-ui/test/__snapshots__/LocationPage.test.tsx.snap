// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocationPage should render LocationPage after click on the button Change 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <h6
      class="MuiTypography-root MuiTypography-h6 change-location css-2ulfj5-MuiTypography-root"
    >
      Change location
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeLarge MuiButton-containedSizeLarge change-button css-mkazzp-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Change
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </h6>
  </div>
</DocumentFragment>
`;

exports[`LocationPage should render LocationPage without opened modal window 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <h6
      class="MuiTypography-root MuiTypography-h6 change-location css-2ulfj5-MuiTypography-root"
    >
      Change location
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeLarge MuiButton-containedSizeLarge change-button css-mkazzp-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Change
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </h6>
  </div>
</DocumentFragment>
`;

exports[`LocationPage should render modal with inputs and active button after typing correct text 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <h6
      class="MuiTypography-root MuiTypography-h6 change-location css-2ulfj5-MuiTypography-root"
    >
      Change location
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeLarge MuiButton-containedSizeLarge change-button css-mkazzp-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Change
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        >
          <span
            class="css-y4cjyz-MuiTouchRipple-ripple MuiTouchRipple-ripple MuiTouchRipple-rippleVisible"
            style="width: 2.8284271247461903px; height: 2.8284271247461903px; top: -1.4142135623730951px; left: -1.4142135623730951px;"
          >
            <span
              class="MuiTouchRipple-child MuiTouchRipple-childLeaving"
            />
          </span>
        </span>
      </button>
    </h6>
  </div>
</DocumentFragment>
`;

exports[`LocationPage should render modal with inputs and active button after typing correct text and show weather forecast 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <h6
      class="MuiTypography-root MuiTypography-h6 change-location css-2ulfj5-MuiTypography-root"
    >
      Change location
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeLarge MuiButton-containedSizeLarge change-button css-mkazzp-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Change
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        >
          <span
            class="css-y4cjyz-MuiTouchRipple-ripple MuiTouchRipple-ripple MuiTouchRipple-rippleVisible"
            style="width: 2.8284271247461903px; height: 2.8284271247461903px; top: -1.4142135623730951px; left: -1.4142135623730951px;"
          >
            <span
              class="MuiTouchRipple-child MuiTouchRipple-childLeaving"
            />
          </span>
        </span>
      </button>
    </h6>
  </div>
</DocumentFragment>
`;

exports[`LocationPage should render modal with inputs and disabled button after typing incorrect text 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <h6
      class="MuiTypography-root MuiTypography-h6 change-location css-2ulfj5-MuiTypography-root"
    >
      Change location
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeLarge MuiButton-containedSizeLarge change-button css-mkazzp-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Change
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        >
          <span
            class="css-y4cjyz-MuiTouchRipple-ripple MuiTouchRipple-ripple MuiTouchRipple-rippleVisible"
            style="width: 2.8284271247461903px; height: 2.8284271247461903px; top: -1.4142135623730951px; left: -1.4142135623730951px;"
          >
            <span
              class="MuiTouchRipple-child MuiTouchRipple-childLeaving"
            />
          </span>
        </span>
      </button>
    </h6>
  </div>
</DocumentFragment>
`;
